internal/events/blockchain_event.go

```diff
@@ -99,32 +99,51 @@ func (em *eventManager) emitBlockchainEventMetric(event *blockchain.Event) {
 	}
 }
 
-func (em *eventManager) BlockchainEvent(event *blockchain.EventWithSubscription) error {
+func (em *eventManager) BlockchainEventBatch(batch []*blockchain.EventToDispatch) error {
 	return em.retry.Do(em.ctx, "persist blockchain event", func(attempt int) (bool, error) {
-		err := em.database.RunAsGroup(em.ctx, func(ctx context.Context) error {
-			listener, err := em.getChainListenerByProtocolIDCached(ctx, event.Subscription)
-			if err != nil {
-				return err
+		return true, em.database.RunAsGroup(em.ctx, func(ctx context.Context) error {
+			for _, event := range batch {
+				switch event.Type {
+				case blockchain.EventTypeForListener:
+					if err := em.handleBlockchainEventForListener(ctx, event.ForListener); err != nil {
+						return err
+					}
+				case blockchain.EventTypeBatchPinComplete:
+					if err := em.handleBlockchainBatchPinEvent(ctx, event.BatchPinComplete); err != nil {
+						return err
+					}
+				case blockchain.EventTypeNetworkAction:
```

Comment:
@peterbroadhurst I spent some time trying to figure out why the test is failing for sqllite3. After reading the code and test. The test was failing on waiting for the NetworkAction event `firefly:terminate` but didn't get it.

Looking at the changes made in this PR, the old logic for emitting network action event was not under "em.database.RunAsGroup" but the new logic does. Could this be the issue for the failing test?

I did notice the same test is working fine with postgres run

Reply:
Possible we've ended up with a nested `RunAsGroup` - I shall dig into it...

Reply:
@peterbroadhurst that explains why postgres worked. Thanks for sharing the details here. 🙏